Add initial specification in Quint (#135)
matter-labs/era-consensus: spec/protocol-spec/README.md

have done with it. This Quint specification was prepared by Igor Konnov and
Denis Kolegov (Matter Labs) under a contract with Matter Labs.

We present the experiments with randomized simulation and bounded model
checking. An inductive proof of safety is work in progress.

## 1. Protocol configurations

We are checking the protocol specification for several configurations:

 - [n6f1b0](./n6f1b0.qnt) is the protocol instance for `N=6`, `F=1`, and no
 faulty replicas. All six replicas are correct, though some of them may be slow.
 This instance must be safe, e.g., `agreement_inv` must not be violated.

 - [n7f1b0](./n7f1b0.qnt) is the protocol instance for `N=7`, `F=1`, and no
 faulty replicas. All seven replicas are correct, though some of them may be slow.
 This instance lets us test whether more correct replicas may introduce issues.

 - [n6f1b1](./n6f1b1.qnt) is the protocol instance for `N=6`, `F=1`, and one
 Byzantine replica. Five correct replicas should be able to tolerate one
 Byzantine replica. This instance must be safe, e.g., `agreement_inv` must not
 be violated.
 
 - [n6f1b2](./n6f1b2.qnt) is the protocol instance for `N=6`, `F=1`, and two

>  Byzantines replica. Two Byzantine replicas have the critical mass to partition

Reviewer: ```suggestion
 Byzantine replicas. Two Byzantine replicas have the critical mass to partition
```

 the correct replicas. Some invariants such as `agreement_inv` are violated in
 this instance.
 
 - [n6f1b3](./n6f1b3.qnt) is the protocol instance for `N=6`, `F=1`, and three

>  Byzantines replica. Many invariants are violated in this instance very quickly.

Reviewer: ```suggestion
 Byzantines replicas. Many invariants are violated in this instance very quickly.
```

|11| `two_chained_blocks_example` | Produce an execution, where a correct replica finalizes two blocks |

There are two ways to find examples, which we consider below.

**Running the randomized simulator.** With this command, the tool is simply replacing non-deterministic choices such as `any` and `oneOf` with random choice. For simple invariants, the randomized simulator produces examples very quickly:

  ```
  $ quint run --invariant=phase_commit_example ./experiments/n6f1b0.qnt
  ```

**Running randomized symbolic execution.** With this command, the tool is randomly choosing
actions in `any {...}` but keeps the choice of data in `oneOf` symbolic. It further offloads the task of finding an execution to the symbolic model checker called [Apalache][]. This is how we run symbolic
execution:

  ```
  $ quint verify --random-transitions=true --invariant=phase_commit_example ./experiments/n6f1b0.qnt
  ```

## 3. State invariants

In the course of writing the specification, we have developed a number of state
invariants. By simulating and model checking the specification against these

> invariants increases our confidence in the correctness of the protocol.

Reviewer: ```suggestion
invariants, we have increased our confidence in the correctness of the protocol.
```

|12| `one_high_vote_in_timeout_qc_inv` | A single `TimeoutQC` does not let us form two high votes |
|13| `one_commit_quorum_inv` | It should not be possible to build two quorums for two different block hashes in the same view |
|14| `all_replicas_high_commit_qc_inv` | Integrity of high `CommitQC`'s stored by correct replicas |
|15| `all_replicas_high_timeout_qc_inv` | Integrity of high `TimeoutQC`'s stored by correct replicas |
|16| `msgs_signed_timeout_inv` | Correct replicas do not send malformed timeout votes |
|17| `msgs_signed_commit_inv` | Correct replicas do not send malformed commit votes |
|18| `all_replicas_high_vote_inv` | Correct replicas do not store malformed high votes |
|19| `timeout_high_vote_is_highest_inv` | The high vote is the highest among the votes sent by a replica |

## 4. Checking state invariants

**Randomized simulator.** Similar to finding examples, we run the randomized simulator
to check correctness of the protocol against state invariants. For example:

```sh
quint run --max-steps=20 --max-samples=100000 --invariant=no_commit_equivocation_inv n6f1b3.qnt
```

> By changing `--max-steps`, we give the simulator how deep the executions should be, e.g., up to 20 steps. By changing `--max-samples`, we give the simulator the number of executions to try, e.g., 100,000.

Reviewer: ```suggestion
By changing `--max-steps`, we tell the simulator how deep the executions should be, e.g., up to 20 steps. By changing `--max-samples`, we give the simulator the number of executions to try, e.g., 100,000.
```
